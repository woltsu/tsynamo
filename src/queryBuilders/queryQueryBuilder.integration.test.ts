import { DDB, TEST_DATA } from "../../test/testFixture";
import { getDDBClientFor, startDDBTestContainer } from "../../test/testUtil";
import { Tsynamo } from "./../index";

describe("QueryQueryBuilder", () => {
  let tsynamoClient: Tsynamo<DDB>;

  beforeAll(async () => {
    const testContainer = await startDDBTestContainer();

    tsynamoClient = new Tsynamo<DDB>({
      ddbClient: await getDDBClientFor(testContainer),
    });
  });

  it("handles a query with a simple KeyCondition", async () => {
    const data = await tsynamoClient
      .query("myTable")
      .keyCondition("userId", "=", TEST_DATA[0].userId)
      .execute();

    expect(data?.length).toBe(5);
    expect(data).toMatchSnapshot();
  });

  it("handles a KeyCondition with BETWEEN expression", async () => {
    const data = await tsynamoClient
      .query("myTable")
      .keyCondition("userId", "=", "123")
      .keyCondition("dataTimestamp", "BETWEEN", 150, 500)
      .execute();

    expect(data).toMatchSnapshot();
  });

  it("handles a KeyCondition with begins_with function", async () => {
    let data = await tsynamoClient
      .query("myOtherTable")
      .keyCondition("userId", "=", "123")
      .keyCondition("stringTimestamp", "begins_with", "11")
      .execute();

    expect(data).toMatchSnapshot();
  });

  it("handles a query with a FilterExpression", async () => {
    const data = await tsynamoClient
      .query("myTable")
      .keyCondition("userId", "=", TEST_DATA[0].userId)
      .filterExpression("someBoolean", "=", true)
      .filterExpression("someBoolean", "=", true)
      .execute();

    expect(data?.length).toBe(2);
    expect(data).toMatchSnapshot();
  });

  it("handles a query with multiple expressions", async () => {
    const data = await tsynamoClient
      .query("myTable")
      .keyCondition("userId", "=", "123")
      .keyCondition("dataTimestamp", "<", 888)
      .filterExpression("someBoolean", "=", true)
      .filterExpression("somethingElse", "<", 0)
      .execute();

    expect(data).toMatchSnapshot();
  });

  it("handles a query with a nested FilterExpression", async () => {
    const data = await tsynamoClient
      .query("myTable")
      .keyCondition("userId", "=", "123")
      .keyCondition("dataTimestamp", "<", 888)
      .filterExpression("somethingElse", "<", 2)
      .orFilterExpression((qb) =>
        qb
          .expression("someBoolean", "=", true)
          .expression("somethingElse", "=", 2)
      )
      .execute();

    expect(data).toMatchSnapshot();
  });

  it("handles a query with a NOT FilterExpression", async () => {
    let data = await tsynamoClient
      .query("myTable")
      .keyCondition("userId", "=", "123")
      .filterExpression("NOT", (qb) => qb.expression("someBoolean", "=", true))
      .execute();

    expect(data).toMatchSnapshot();

    data = await tsynamoClient
      .query("myTable")
      .keyCondition("userId", "=", "123")
      .filterExpression("someBoolean", "=", true)
      .orFilterExpression("NOT", (qb) => qb.expression("somethingElse", "=", 0))
      .execute();

    expect(data).toMatchSnapshot();
  });

  it("handles a query with a BETWEEN FilterExpression", async () => {
    let data = await tsynamoClient
      .query("myTable")
      .keyCondition("userId", "=", "123")
      .filterExpression("somethingElse", "BETWEEN", 0, 10)
      .execute();

    expect(data).toMatchSnapshot();

    data = await tsynamoClient
      .query("myTable")
      .keyCondition("userId", "=", "123")
      .filterExpression("someBoolean", "=", true)
      .orFilterExpression("somethingElse", "BETWEEN", 9, 10)
      .execute();

    expect(data).toMatchSnapshot();
  });

  it("handles conditions on nested keys", async () => {
    const data = await tsynamoClient
      .query("myTable")
      .keyCondition("userId", "=", "123")
      .filterExpression("nested.nestedBoolean", "=", true)
      .execute();

    expect(data).toMatchSnapshot();
  });

  it("handles a FilterExpression that uses attribute_exists and attribute_not_exists", async () => {
    const data = await tsynamoClient
      .query("myTable")
      .keyCondition("userId", "=", "123")
      .filterExpression("nested.nestedString", "attribute_exists")
      .execute();

    expect(data).toMatchSnapshot();

    const data2 = await tsynamoClient
      .query("myTable")
      .keyCondition("userId", "=", "123")
      .filterExpression("nested.nestedString", "attribute_not_exists")
      .attributes(["userId", "nested"])
      .execute();

    expect(data2).toMatchSnapshot();
  });

  it("handles a FilterExpression that uses begins_with", async () => {
    let data = await tsynamoClient
      .query("myTable")
      .keyCondition("userId", "=", "123")
      .filterExpression("nested.nestedString", "begins_with", "k")
      .execute();

    expect(data).toMatchSnapshot();

    data = await tsynamoClient
      .query("myTable")
      .keyCondition("userId", "=", "123")
      .filterExpression("nested.nestedString", "begins_with", "ke")
      .execute();

    expect(data).toMatchSnapshot();
  });

  it("handles a FilterExpression that takes attributes from cats array", async () => {
    let data = await tsynamoClient
      .query("myOtherTable")
      .keyCondition("userId", "=", "123")
      .filterExpression("cats", "attribute_exists")
      .attributes(["cats[0].age"])
      .execute();

    expect(data).toMatchSnapshot();
  });

  it("handles a FilterExpression that uses CONTAINS function", async () => {
    const data = await tsynamoClient
      .query("myTable")
      .keyCondition("userId", "=", "313")
      .filterExpression("tags", "contains", "testTag")
      .execute();

    expect(data).toMatchSnapshot();
  });

  it("handles a query with limit", async () => {
    const data = await tsynamoClient
      .query("myTable")
      .keyCondition("userId", "=", "123")
      .limit(2)
      .execute();

    expect(data).toHaveLength(2);
    expect(data).toMatchSnapshot();
  })

  it("handles a query with index", async () => {
    const data = await tsynamoClient
      .query("myTable")
      .keyCondition("userId", "=", "123")
      .index("myIndex")
      .execute();

    expect(data).toMatchSnapshot();
  })

  it("fails a query with non-existent index", async () => {
    await expect(tsynamoClient
      .query("myTable")
      .keyCondition("userId", "=", "123")
      .index("nonExistentIndex")
      .execute()
    ).rejects.toThrowErrorMatchingSnapshot();
  })
});
