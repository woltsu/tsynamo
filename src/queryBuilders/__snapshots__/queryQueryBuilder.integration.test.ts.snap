// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`QueryQueryBuilder > fails a query with non-existent index 1`] = `[ValidationException: The table does not have the specified index: nonExistentIndex]`;

exports[`QueryQueryBuilder > handles a FilterExpression that takes attributes from cats array 1`] = `
[
  {
    "cats": [
      {
        "age": 42,
      },
    ],
  },
]
`;

exports[`QueryQueryBuilder > handles a FilterExpression that uses CONTAINS function 1`] = `
[
  {
    "dataTimestamp": 996,
    "someBoolean": false,
    "somethingElse": -9,
    "tags": [
      "testTag",
    ],
    "userId": "313",
  },
]
`;

exports[`QueryQueryBuilder > handles a FilterExpression that uses attribute_exists and attribute_not_exists 1`] = `
[
  {
    "dataTimestamp": 111,
    "nested": {
      "nestedBoolean": false,
      "nestedString": "koy",
    },
    "someBoolean": true,
    "somethingElse": -5,
    "userId": "123",
  },
  {
    "dataTimestamp": 996,
    "nested": {
      "nestedBoolean": true,
      "nestedNested": {
        "nestedNestedBoolean": true,
      },
      "nestedString": "llol",
    },
    "someBoolean": false,
    "somethingElse": -9,
    "userId": "123",
  },
  {
    "dataTimestamp": 999,
    "nested": {
      "nestedBoolean": true,
      "nestedString": "key",
    },
    "someBoolean": false,
    "somethingElse": 0,
    "userId": "123",
  },
]
`;

exports[`QueryQueryBuilder > handles a FilterExpression that uses attribute_exists and attribute_not_exists 2`] = `
[
  {
    "userId": "123",
  },
  {
    "userId": "123",
  },
]
`;

exports[`QueryQueryBuilder > handles a FilterExpression that uses begins_with 1`] = `
[
  {
    "dataTimestamp": 111,
    "nested": {
      "nestedBoolean": false,
      "nestedString": "koy",
    },
    "someBoolean": true,
    "somethingElse": -5,
    "userId": "123",
  },
  {
    "dataTimestamp": 999,
    "nested": {
      "nestedBoolean": true,
      "nestedString": "key",
    },
    "someBoolean": false,
    "somethingElse": 0,
    "userId": "123",
  },
]
`;

exports[`QueryQueryBuilder > handles a FilterExpression that uses begins_with 2`] = `
[
  {
    "dataTimestamp": 999,
    "nested": {
      "nestedBoolean": true,
      "nestedString": "key",
    },
    "someBoolean": false,
    "somethingElse": 0,
    "userId": "123",
  },
]
`;

exports[`QueryQueryBuilder > handles a KeyCondition with BETWEEN expression 1`] = `
[
  {
    "dataTimestamp": 222,
    "someBoolean": true,
    "somethingElse": 2,
    "userId": "123",
  },
  {
    "dataTimestamp": 333,
    "someBoolean": false,
    "somethingElse": 10,
    "userId": "123",
  },
]
`;

exports[`QueryQueryBuilder > handles a KeyCondition with begins_with function 1`] = `
[
  {
    "someBoolean": true,
    "somethingElse": -5,
    "stringTimestamp": "111",
    "userId": "123",
  },
]
`;

exports[`QueryQueryBuilder > handles a query with a BETWEEN FilterExpression 1`] = `
[
  {
    "dataTimestamp": 222,
    "someBoolean": true,
    "somethingElse": 2,
    "userId": "123",
  },
  {
    "dataTimestamp": 333,
    "someBoolean": false,
    "somethingElse": 10,
    "userId": "123",
  },
  {
    "dataTimestamp": 999,
    "nested": {
      "nestedBoolean": true,
      "nestedString": "key",
    },
    "someBoolean": false,
    "somethingElse": 0,
    "userId": "123",
  },
]
`;

exports[`QueryQueryBuilder > handles a query with a BETWEEN FilterExpression 2`] = `
[
  {
    "dataTimestamp": 111,
    "nested": {
      "nestedBoolean": false,
      "nestedString": "koy",
    },
    "someBoolean": true,
    "somethingElse": -5,
    "userId": "123",
  },
  {
    "dataTimestamp": 222,
    "someBoolean": true,
    "somethingElse": 2,
    "userId": "123",
  },
  {
    "dataTimestamp": 333,
    "someBoolean": false,
    "somethingElse": 10,
    "userId": "123",
  },
]
`;

exports[`QueryQueryBuilder > handles a query with a FilterExpression 1`] = `
[
  {
    "dataTimestamp": 111,
    "nested": {
      "nestedBoolean": false,
      "nestedString": "koy",
    },
    "someBoolean": true,
    "somethingElse": -5,
    "userId": "123",
  },
  {
    "dataTimestamp": 222,
    "someBoolean": true,
    "somethingElse": 2,
    "userId": "123",
  },
]
`;

exports[`QueryQueryBuilder > handles a query with a NOT FilterExpression 1`] = `
[
  {
    "dataTimestamp": 333,
    "someBoolean": false,
    "somethingElse": 10,
    "userId": "123",
  },
  {
    "dataTimestamp": 996,
    "nested": {
      "nestedBoolean": true,
      "nestedNested": {
        "nestedNestedBoolean": true,
      },
      "nestedString": "llol",
    },
    "someBoolean": false,
    "somethingElse": -9,
    "userId": "123",
  },
  {
    "dataTimestamp": 999,
    "nested": {
      "nestedBoolean": true,
      "nestedString": "key",
    },
    "someBoolean": false,
    "somethingElse": 0,
    "userId": "123",
  },
]
`;

exports[`QueryQueryBuilder > handles a query with a NOT FilterExpression 2`] = `
[
  {
    "dataTimestamp": 111,
    "nested": {
      "nestedBoolean": false,
      "nestedString": "koy",
    },
    "someBoolean": true,
    "somethingElse": -5,
    "userId": "123",
  },
  {
    "dataTimestamp": 222,
    "someBoolean": true,
    "somethingElse": 2,
    "userId": "123",
  },
  {
    "dataTimestamp": 333,
    "someBoolean": false,
    "somethingElse": 10,
    "userId": "123",
  },
  {
    "dataTimestamp": 996,
    "nested": {
      "nestedBoolean": true,
      "nestedNested": {
        "nestedNestedBoolean": true,
      },
      "nestedString": "llol",
    },
    "someBoolean": false,
    "somethingElse": -9,
    "userId": "123",
  },
]
`;

exports[`QueryQueryBuilder > handles a query with a nested FilterExpression 1`] = `
[
  {
    "dataTimestamp": 111,
    "nested": {
      "nestedBoolean": false,
      "nestedString": "koy",
    },
    "someBoolean": true,
    "somethingElse": -5,
    "userId": "123",
  },
  {
    "dataTimestamp": 222,
    "someBoolean": true,
    "somethingElse": 2,
    "userId": "123",
  },
]
`;

exports[`QueryQueryBuilder > handles a query with a simple KeyCondition 1`] = `
[
  {
    "dataTimestamp": 111,
    "nested": {
      "nestedBoolean": false,
      "nestedString": "koy",
    },
    "someBoolean": true,
    "somethingElse": -5,
    "userId": "123",
  },
  {
    "dataTimestamp": 222,
    "someBoolean": true,
    "somethingElse": 2,
    "userId": "123",
  },
  {
    "dataTimestamp": 333,
    "someBoolean": false,
    "somethingElse": 10,
    "userId": "123",
  },
  {
    "dataTimestamp": 996,
    "nested": {
      "nestedBoolean": true,
      "nestedNested": {
        "nestedNestedBoolean": true,
      },
      "nestedString": "llol",
    },
    "someBoolean": false,
    "somethingElse": -9,
    "userId": "123",
  },
  {
    "dataTimestamp": 999,
    "nested": {
      "nestedBoolean": true,
      "nestedString": "key",
    },
    "someBoolean": false,
    "somethingElse": 0,
    "userId": "123",
  },
]
`;

exports[`QueryQueryBuilder > handles a query with index 1`] = `
[
  {
    "dataTimestamp": 333,
    "someBoolean": false,
    "somethingElse": 10,
    "userId": "123",
  },
  {
    "dataTimestamp": 999,
    "nested": {
      "nestedBoolean": true,
      "nestedString": "key",
    },
    "someBoolean": false,
    "somethingElse": 0,
    "userId": "123",
  },
  {
    "dataTimestamp": 996,
    "nested": {
      "nestedBoolean": true,
      "nestedNested": {
        "nestedNestedBoolean": true,
      },
      "nestedString": "llol",
    },
    "someBoolean": false,
    "somethingElse": -9,
    "userId": "123",
  },
  {
    "dataTimestamp": 222,
    "someBoolean": true,
    "somethingElse": 2,
    "userId": "123",
  },
  {
    "dataTimestamp": 111,
    "nested": {
      "nestedBoolean": false,
      "nestedString": "koy",
    },
    "someBoolean": true,
    "somethingElse": -5,
    "userId": "123",
  },
]
`;

exports[`QueryQueryBuilder > handles a query with limit 1`] = `
[
  {
    "dataTimestamp": 111,
    "nested": {
      "nestedBoolean": false,
      "nestedString": "koy",
    },
    "someBoolean": true,
    "somethingElse": -5,
    "userId": "123",
  },
  {
    "dataTimestamp": 222,
    "someBoolean": true,
    "somethingElse": 2,
    "userId": "123",
  },
]
`;

exports[`QueryQueryBuilder > handles a query with multiple expressions 1`] = `
[
  {
    "dataTimestamp": 111,
    "nested": {
      "nestedBoolean": false,
      "nestedString": "koy",
    },
    "someBoolean": true,
    "somethingElse": -5,
    "userId": "123",
  },
]
`;

exports[`QueryQueryBuilder > handles conditions on nested keys 1`] = `
[
  {
    "dataTimestamp": 996,
    "nested": {
      "nestedBoolean": true,
      "nestedNested": {
        "nestedNestedBoolean": true,
      },
      "nestedString": "llol",
    },
    "someBoolean": false,
    "somethingElse": -9,
    "userId": "123",
  },
  {
    "dataTimestamp": 999,
    "nested": {
      "nestedBoolean": true,
      "nestedString": "key",
    },
    "someBoolean": false,
    "somethingElse": 0,
    "userId": "123",
  },
]
`;
